spec/validations/requests_spec.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+def request_function(style, address)
+  case style
+  when 'ZAF'
+    "function request() { client.request('#{address}') }"
+  when 'jQuery'
+    "function request() { $.get('#{address}') }"
+  when 'XMLHttpRequest'
+    "function request() { xhr.open('get', '#{address}') }"
+  when 'fetch'
+    "function request() { fetch('#{address}') }"
+  end
+end
+
+request_function_styles = ['ZAF', 'jQuery', 'XMLHttpRequest', 'fetch']
+
+shared_examples 'an insecure request' do |function_style|
+  let(:markup) { request_function(function_style, 'http://foo.com') }
+
+  it "and raise a warning inside #{function_style} style requests" do
+    # TODO: update with translated string
+    warning = 'Possible insecure HTTP request to http://foo.com in assets/app.js. Consider instead using the HTTPS protocol.'
+    errors = subject.call(package)
+    expect(package.warnings).to include(warning)
+    expect(errors).to be_empty
+  end
+end
+
+blocked_requests = [
+  {
+    type: 'private',
+    example_address: 'https://192.168.0.1',
+    # TODO: update with translated string
+    error: 'Possible request to private network address 192.168.0.1 in assets/app.js'
+  },
+  {
+    type: 'loopback',
+    example_address: 'https://127.0.0.1',
+    # TODO: update with translated string
+    error: 'Possible request to a loopback network address 127.0.0.1 in assets/app.js'
+  },
+  { type: 'link-local',
+    example_address: 'https://169.254.0.1',
```

Comment:
nit: worth testing that `http://` link-local addresses get a warning and an error?

Reply:
Nah, it wouldn't. The way ZAS is wired up, errors trump warnings and warnings are only displayed in the absence of errors. Not great since developers might have to do multiple validation passes but I think that's a change outside the scope of this PR

Reply:
Ahh, makes sense.